lib/events.js

```diff
@@ -512,3 +514,76 @@ function once(emitter, name) {
     emitter.once(name, eventListener);
   });
 }
+
+const AsyncIteratorPrototype = Object.getPrototypeOf(
+  Object.getPrototypeOf(async function* () {}).prototype);
+
+function createIterResult(value, done) {
+  return { value, done };
+}
+
+function on(emitter, event) {
+  const unconsumedEvents = [];
+  const unconsumedPromises = [];
+  let error = null;
+
+  const iterator = Object.setPrototypeOf({
+    next() {
+      if (error) {
+        return Promise.reject(error);
+      }
+
+      const value = unconsumedEvents.shift();
+      if (value) {
+        return Promise.resolve(createIterResult(value, false));
+      }
+
+      return new Promise(function(resolve, reject) {
+        unconsumedPromises.push({ resolve, reject });
+      });
+    },
+
+    return() {
+      emitter.removeListener(event, eventHandler);
+      emitter.removeListener('error', errorHandler);
+
+      return Promise.resolve(createIterResult(undefined, true));
+    },
+
+    throw(err) {
+      assert(err);
+      error = err;
+      emitter.removeListener(event, eventHandler);
+      emitter.removeListener('error', errorHandler);
+    },
+
+    [Symbol.asyncIterator]() {
+      return this;
+    }
+  }, AsyncIteratorPrototype);
```

Comment:
are we doing it this way because it's faster than using `async function * () {}` directly? Or it just way less ergononmic/maintainable to us an async generator for this case?

Reply:
What do you mean? How could this be written with an async generator?

The only way I think of will resubscribe which can drop events